### PR TITLE
kube-proxy: internal config: add Linux and Windows section

### PR DESCRIPTION
**Pull request by aroradaman**
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
Introduce Linux and Windows section for internal configuration of kube-proxy adhering to the v1alpha2 version specifications as detailed in https://kep.k8s.io/784. 
This also introduces WindowsRunAsService to the v1alpha1 configuration.
#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
kube-proxy Windows service control manager integration(--windows-service) is now configurable in v1alpha1 component configuration via `WindowsRunAsService` field
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: pkg/proxy/apis/config/fuzzer/fuzzer.go [hunk 1]
```diff
@@ -51,6 +51,7 @@ func Funcs(codecs runtimeserializer.CodecFactory) []interface{} {
 			if obj.Logging.Format == "" {
 				obj.Logging.Format = "text"
 			}
+			obj.Windows = kubeproxyconfig.KubeProxyWindowsConfiguration{}
```
Comment (danwinship) [inline pkg/proxy/apis/config/fuzzer/fuzzer.go:54]: You don't need to do that, and in fact you don't _want_ to do that; the `c.FuzzNoCustom(obj)` will already have fuzzed all of the scalar-typed fields (including `obj.Windows.RunAsService`), and this line will effectively undo that and force `RunAsService` to always be `false`.

Or did you add this to avoid unit test problems? It's possible that we're just going to have to add the new fields to the `v1alpha` config too...

File: pkg/proxy/apis/config/types.go [hunk 2]
```diff
@@ -22,6 +22,13 @@ import (
 	logsapi "k8s.io/component-base/logs/api/v1"
 )
 
+// KubeProxyWindowsConfiguration contains windows-platform-related configuration details for the
+// Kubernetes proxy server that aren't specific to a particular backend
+type KubeProxyWindowsConfiguration struct {
+	// runAsService, if true, enables windows service control manager API integration.
```
Comment (danwinship) [inline pkg/proxy/apis/config/types.go:28]:
```suggestion
	// runAsService, if true, enables Windows service control manager API integration.
```

likewise in the other comments

File: pkg/proxy/apis/config/v1alpha1/conversion.go [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+	"k8s.io/kube-proxy/config/v1alpha1"
+	"k8s.io/kubernetes/pkg/proxy/apis/config"
+)
+
+// Convert_config_KubeProxyConfiguration_To_v1alpha1_KubeProxyConfiguration is defined here, because public conversion is not auto-generated due to existing warnings.
+func Convert_config_KubeProxyConfiguration_To_v1alpha1_KubeProxyConfiguration(in *config.KubeProxyConfiguration, out *v1alpha1.KubeProxyConfiguration, scope conversion.Scope) error {
+	if err := autoConvert_config_KubeProxyConfiguration_To_v1alpha1_KubeProxyConfiguration(in, out, scope); err != nil {
+		return err
+	}
+	return nil
```
Comment (danwinship) [inline pkg/proxy/apis/config/v1alpha1/conversion.go:30]: I assume you did it this way rather than just `return autoConvert_config_KubeProxyConfiguration_To_v1alpha1_KubeProxyConfiguration(in, out, scope)` because other stuff is going to be added to this function in future commits?

Comment (aroradaman) [reply pkg/proxy/apis/config/v1alpha1/conversion.go:30]: Yes, this was intentional.

File: pkg/proxy/apis/config/fuzzer/fuzzer.go (hunk 1 above)
Comment (aroradaman) [reply pkg/proxy/apis/config/fuzzer/fuzzer.go:54]: Yes, the unit test fails if I don't add that.
Should I simply add `WindowsRunAsService` to `v1alpha1.KubeProxyConfiguration`?

File: staging/src/k8s.io/kube-proxy/config/v1alpha1/types.go [hunk 4]
```diff
@@ -243,6 +243,9 @@ type KubeProxyConfiguration struct {
 
 	// portRange was previously used to configure the userspace proxy, but is now unused.
 	PortRange string `json:"portRange"`
+
+	// windowsRunAsService, if true, enables Windows service control manager API integration.
+	WindowsRunAsService bool `json:"windowsRunAsService"`
```
Comment (danwinship) [inline staging/src/k8s.io/kube-proxy/config/v1alpha1/types.go:248]: If you add `omitempty` then it won't show up in the output by default and you can revert the changes to the unit test data.

The alternative to this field would be to just add the whole `KubeProxyWindowsConfiguration` to v1alpha1. But I guess that would be weird since we won't need to add `KubeProxyLinuxConfiguration`. So this seems fine.

File: pkg/proxy/apis/config/types.go [hunk 5]
```diff
@@ -165,6 +172,9 @@ type DetectLocalConfiguration struct {
 type KubeProxyConfiguration struct {
 	metav1.TypeMeta
 
+	// windows contains windows-related configuration options.
```
Comment (danwinship) [inline pkg/proxy/apis/config/types.go:175]:
```suggestion
	// windows contains Windows-related configuration options.
```

**Comment by aroradaman** [comment]
/test pull-kubernetes-node-e2e-containerd 

_Pod scheduling timeout_

File: pkg/proxy/apis/config/v1alpha1/conversion.go [hunk 6]
```diff
@@ -28,6 +28,10 @@ func Convert_config_KubeProxyConfiguration_To_v1alpha1_KubeProxyConfiguration(in
 		return err
 	}
 	out.WindowsRunAsService = in.Windows.RunAsService
+	out.Conntrack = v1alpha1.KubeProxyConntrackConfiguration(in.Linux.Conntrack)
+	out.OOMScoreAdj = in.Linux.OOMScoreAdj
+	out.NFTables.MasqueradeAll = in.Linux.MasqueradeAll
+	out.IPTables.MasqueradeAll = in.Linux.MasqueradeAll
```
Comment (danwinship) [inline pkg/proxy/apis/config/v1alpha1/conversion.go:34]: I'd probably make this conditional on `in.Mode` like with the other conversion func. (Ideally the mode-specific configs should all have default values when not using that mode...)

**Comment by danwinship** [comment]
/lgtm
/approve

/assign @thockin
for API approval (adds a new field to the `v1alpha1` kube-proxy API corresponding to an existing command-line option / future `v1alpha2` config option)

This is part of #121830, which is reorganizing the internal version of `KubeProxyConfiguration` to make more sense, in preparation for the `v1alpha2` (and some day `v1`) API. This part just splits out the `Linux` and `Windows` sections, as detailed more in [the KEP](https://github.com/kubernetes/enhancements/tree/master/keps/sig-network/784-kube-proxy-component-config#design-details). Needs API approval because of the new `v1alpha1` field (which, to be clear, we don't really intend for anyone to use; we just wanted it in `v1alpha2` and ran into round-tripping problems if we didn't add it to `v1alpha1` too).

**Comment by thockin** [comment]
@danwinship you should add yourself in staging/src/k8s.io/kube-proxy/config/OWNERS

**Comment by thockin** [comment]
I am not sure why we have all the component-config schemas in pkg/generated/openapi/zz_generated.openapi.go -- is it intentional or accidental?  Maybe @jpbetz or @liggitt knows?

/approve
